### DISCUSSION TITLE
Use nested SASS classes with css modules

### DISCUSSION THREAD

**Discussion by martonlederer** [vercel/next.js#19417] [Help] [answered]
This is my css/sass module:
```sass
.landingProductSmall, .landingProductLarge
  cursor: pointer
  position: relative

  .dots
    position: absolute
    background-image: url(/background.svg)
    background-size: 3em
    background-repeat: repeat
    top: 0
    left: 0
    right: 0
    bottom: 0

  &:hover
    .dots
      background-size: 3.4em
```
The component:
```tsx
<div className={styles.landingProductSmall}>
  <div className="dots"></div>
</div>
```
It does not apply the `.dots` style to the children of `.landingProductSmall` / `.landingProductLarge`.
Any suggestions?

**Comment by jamesmosier** [comment]
You cannot nest classes with SASS (or CSS) modules like you have it. You would have to also ensure `.dots` is a SASS module so that when compiled it is able to understand and generate the correct class names:

```js
<div className={styles.landingProductSmall}>
  <div className{styles.dots}></div>
</div>
```

---

otherwise you could do:

```js
<div className={styles.landingProductSmall}>
  <div></div>
</div>
```

```sass
.landingProductSmall
  cursor: pointer
  
  div 
    position: absolute
```

to target any child `div`. But otherwise trying to target a nested class will not work.

**Reply by bohdanyurov-gl** [reply]
@jamesmosier Unfortunately it wouldn't work in every case some class is set in 3rd party component.

Simple example - react-bootstrap. It sets "active" that we can't modify and can't target using module css.

Btw, why are nested css classes even parsed? Is there any way to disable this behavior?

**Reply by pcjmfranken** [reply]
> Simple example - react-bootstrap. It sets "active" that we can't modify and can't target using module css

I'm not familiar with react-bootstrap, but I cant imagine they've implemented any active-state in such a way that it can't be targeted and overridden by your own custom styling. I'd be happy to take a look if you provide an example!

As for the nested css classes, you could look into global scoping[1] for individual classes or sections. Alternatively, if a particular class should be globally available, perhaps you could define it in a global (non-module) CSS/SASS file instead.

[1] https://github.com/css-modules/css-modules#exceptions

**Reply by bohdanyurov-gl** [reply]
Well, unfortunately the "active" class is hardcoded in bootstrap dropdown, slider and other components. I think it might be just a bootstrap convention. It's not just Bootstrap. font-awesome also has some conventions on classes. Basically every react module that provides react mappings for css-related libraries can't be used with component-level css/scss due to dependency on specific class names that could not be changed (thus it is impossible to provide them with component-level class name).

**UPD** @pcjmfranken thank you a lot! looks like :global() selector is exactly what I've been looking for. Now it's time to go through all other people's questions on StackOverflow as I've seen this being brought up many times.

@martonlederer this is an old question but looks like it should work for your case too.

**Reply by pcjmfranken** [reply]
Cool! Yeah I thought that might have been it. 

If you use a preprocessor that allows nesting you can even global scope entire blocks like this:

```scss
/* styles.module.scss */

.myCustomClass {
  color: hotpink;
  text-decoration: none;
  
  :global {
    &.active {
      font-weight: 700;
    }

    &.active:hover {
      color: deeppink;
      text-decoration: underline;
    }
  }
}
```

Not sure if the syntax is 100% correct and can't currently check, but it should be close. Let me know!

**Comment by martonlederer** [comment, marked as answer]
Returning to this, nowdays i just use `[class~=dots]`

**Reply by aslaker** [reply]
Thank you for circling back on this @martonlederer this saved me another few hours of banging my head against the desk.

**Reply by leobenkel** [reply]
How do you use this ? Sorry I am new to reactjs and nextjs.

**Reply by GianGC1998** [reply]
thanks @martonlederer its save me to override my project 🙇

**Reply by davidtranjs** [reply]
Can you provide any document for your solution ? I'm facing your issue too : (((

**Reply by topched** [reply]
This is a great solution. Helped me override some imported styles from a node module. Thanks!

**Reply by YanivBah** [reply]
too bad it not working for BEM class names like

```scss
.card {
  border: 1px solid black;
  
  &__title {
    font-size: 30px;
    color: red
    &__span {
      font-size: 15px;
      margin-left: 10px;
      }
    }
  
  &__link {
    color: blue;
    &:hover {
    text-decoration: underline;
    }
   }
  
 }
```

Cant back to normal css without nesting :(

**Reply by zachwolf** [reply]
Seems like `:global` is the ticket. IMO the [docs](https://github.com/css-modules/css-modules#exceptions) are confusing here, but it was explained more simply in https://github.com/css-modules/css-modules/issues/188#issuecomment-258892242

**Reply by lgibso34** [reply]
anyone who needs an example:
same as `.test .test2` or `.test{ .test2{} }`

```css
.test{ // name of module
  color:pink;
    [class~="test2"]{ // target child class
       color: blue
    }
}
```

**Reply by ctsstc** [reply]
Be careful using the `[class!~="someClass"]` it will match on "containing" so `[class~="cat"]` would match `.cat` but also `.caterpillar`. I went into more depth here: https://github.com/vercel/next.js/discussions/19417#discussioncomment-1278853 but figured this should be mentioned here since it's marked as the answer.

I highly recommend using `:global` instead. An example can be seen here: https://github.com/vercel/next.js/discussions/19417#discussioncomment-1439398

**Reply by gisheri** [reply]
@ctsstc are you sure? The `~=` operator only matches on exact words within a space separated list (perfect for classes). Not the same as a wildcard match `*=`, so this solution isn't so brittle. I had to double check for myself to be sure, but it seems to hold up. https://codesandbox.io/s/modern-pond-sxv11?file=/src/index.js

https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors
> [attr~=value] Represents elements with an attribute name of attr whose value is a whitespace-separated list of words, one of which is exactly value

**Reply by ctsstc** [reply]
Ah that looks correct, I must have been looking at bad documentation or the wrong one 👍

**Comment by ctsstc** [comment]
It seems there's no documentation around this, and that it's hard to find much information online about this.

I'm just wondering if many people are not using `scss` modules, as this seems to go against what many would expect to happen (standard scss would transpile out to something much different). It's also odd that all class selectors become top-level exports it seems with scss modules. This also seems to defeat the semantic meaning of nesting in the first place, and is confusing what is actually produced or exported. As stated it makes overriding libraries a nightmare or nearly impossible, you may find yourself writing brittle non-descript selectors like `&>div:first-child>span:nth-child(2)` sprinkled with `!important`s to get the job done with overriding.

I honestly can't seem to find if this is a bug or a feature. If it's a feature I would love to better understand the reasoning why scss modules seem to transpile differently than standard scss.

I know this isn't next.js' problem to solve, but I would like to get to the bottom of this and either understand if this is on purpose or if there should be a bug ticket submitted to some other project?

Using attribute contains selectors like `[class~=dots]` may "work" but honestly is a brittle hack, and falls apart as soon as a class name is a substring of another class name.

I'm not sure why it wouldn't just transpile something like:

```scss
.hello {
  .world {
    color: red;
  }
}
```

To something like this: 

```css
.hello_HASH .world {
  color: red;
}
```

or like:

```scss
.something {
  &.active {
    color: red;
  }
}
```

to: 

```css
.something_HASH.active {
  color: red;
}
```

I know this still doesn't necessarily help override selectors, but it gets us closer, as the solution becomes to wrap your overrides in a container.

I wonder if they're concerned that selectors would leak out, or that they don't fully transpile the same so as to avoid override/leakout hacks. I know they catch "non-pure" selectors:
> Syntax error: Selector "body" is not pure (pure selectors must contain at least one local class or id)

**Comment by aestheticsdata** [comment]
same issue here. So what's the point to use scss module if we are forced to use it like a regular css module ?

**Reply by ctsstc** [reply]
>> Redacted; meant to reply to the main thread <<

**Reply by aestheticsdata** [reply]
I'm sorry but I was not talking about :global

**Reply by ctsstc** [reply]
Whoops meant to reply to the main thread not this comment

**Comment by ctsstc** [comment]
Coming back here I'm not sure how I missed `:global` but that's been working great for me. It's just a weird non-standard pain it seems. I thought for a second it would potentially be blocked by css processors because I thought it could pollute the global namespace, but it seems it respects the nesting for building the rules.

Here's an example of something I did recently.
In this example I applied the fixes to the component's div container, but you could also apply them to individual elements/components as well.

Note: from my experience you cannot use do something like this where it applies global to the nested rules unfortunately: `:global { .something {} }`

`component.module.scss`
```javascript
.plyrFix {
  :global .plyr, :global .plyr__video-wrapper {
    height: 100%;
  }
}

.toggleFix {
  button:global.active {
    background-color: red;
    border-color: red;
  }
}
```

`component.js`
```javascript
import styles from './component.module.scss'

export default Component = () => {
  return (
    <div className={`${styles.plyrFix} ${styles.toggleFix}`}>
      Things that need fixing here
    </div>
  )
}
```

**Reply by pcjmfranken** [reply]
As I previously pointed out [0] 😉

[0] https://github.com/vercel/next.js/discussions/19417#discussioncomment-381306

**Comment by pixelkritzel** [comment]
As it turns out you cannest CSS selectors with CSS modules, if you put the ampersand character in the nesting line.

So this

```
.landingProductSmall, .landingProductLarge
  cursor: pointer
  position: relative

  .dots
```

has to become this

```
.landingProductSmall, .landingProductLarge
  cursor: pointer
  position: relative

  &.dots
  
```

**Reply by ghost** [reply]
It doesn't work as well

**Reply by pixelkritzel** [reply]
Works here (after clicking the error message away, idk where it comes from)

https://github.com/pixelkritzel/next-sass-test/blob/main/pages/index.module.sass

**Reply by Oddward** [reply]
Your two examples are two diff style definitions as far as I understand. The first one selects `.landingProductSmall .dots` (descendant) while the second selects `.landingProductSmall.dots` (multiple class names), and they have similar issues in Next.js. 

The ampersand is used when you don't want a space before the nested selector (e.g. :hover, ::before, multiple classes etc)

**Reply by pixelkritzel** [reply]
Sorry, my bad. But as you can see the linked example it works.

**Comment by buzzguy** [comment]
It would be nice to have a “sass modules” where using an ampersand the sass way could be the alternative to using “composes”.
